perf(allocator/vec2): calling `Bump::grow` or `Bump::shrink` at the call site directly instead of calling `realloc`

Dunqing opened the pull request:
This PR aims to remove the `realloc` method, which was copied from `bumpalo`. The `realloc` function is used to call the `shrink` or `grow` method of bumpalo by checking the new size. This is unnecessary because in the call site, we know which method we should call, and the ZST check is also unnecessary because it has been checked earlier in the call site.

Also, this is aligned with the standard library implementation.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10686?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10686** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10686?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10686?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#10675** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10675?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#9855](https://github.com/oxc-project/oxc/pull/9855) <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9855?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#10670** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10670?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10656** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10656?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10654** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10654?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10686).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **May 3, 9:02 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(allocator/vec2): calling `Bump::grow` or `Bump::shrink` at the call site directly instead of calling `realloc` (#10686)

This PR aims to remove the `realloc` method, which was copied from `bumpalo`. The `realloc` function is used to call the `shrink` or `grow` method of bumpalo by checking the new size. This is unnecessary because in the call site, we know which method we should call, and the ZST check is also unnecessary because it has been checked earlier in the call site.

Also, this is aligned with the standard library implementation.
